psydac/linalg/solvers.py

```diff
@@ -1725,3 +1728,261 @@ def solve(self, b, out=None):
+        self._domain = domain
+        self._codomain = codomain
+        self._solver = 'gmres'
+        self._options = {"x0":x0, "tol":tol, "maxiter":maxiter, "verbose":verbose}
+        self._check_options(**self._options) 
+        self._tmps = {key: domain.zeros() for key in ("r", "p", "v", "lv")}
+
+        # Initialize upper Hessenberg matrix
+        self._H = np.zeros((self._options["maxiter"] + 1, self._options["maxiter"]), dtype=A.dtype)
+
+        self._info = None
+        
+
+    def _check_options(self, **kwargs):
+        for key, value in kwargs.items():
+
+            if key == 'x0':
+                if value is not None:
+                    assert isinstance(value, Vector), "x0 must be a Vector or None"
+                    assert value.space == self._codomain, "x0 belongs to the wrong VectorSpace"
+            elif key == 'tol':
+                assert value is not None, "tol may not be None"
+                assert value*0 == 0, "tol must be a real number"
+                assert value > 0, "tol must be positive"
+            elif key == 'maxiter':
+                assert value is not None, "maxiter may not be None"
+                assert isinstance(value, int), "maxiter must be an int"
+                assert value > 0, "maxiter must be positive"
+            elif key == 'verbose':
+                assert value is not None, "verbose may not be None"
+                assert isinstance(value, bool), "verbose must be a bool"
+            else:
+                raise ValueError(f"Key '{key}' not understood. See self._options for allowed keys.")
+
+    def transpose(self, conjugate=False):
+        At = self._A.transpose(conjugate=conjugate)
+        solver = self._solver
+        options = self._options
+        return inverse(At, solver, **options)
+
+    def solve(self, b, out=None):
+        """
+        Generalized minimum residual algorithm for solving linear system Ax=b.
+        Implementation from Wikipedia
+        Info can be accessed using get_info(), see :func:~`basic.InverseLinearOperator.get_info`.
+
+        Parameters
+        ----------
+        b : psydac.linalg.basic.Vector
+            Right-hand-side vector of linear system Ax = b. Individual entries b[i] need
+            not be accessed, but b has 'shape' attribute and provides 'copy()' and
+            'dot(p)' functions (dot(p) is the vector inner product b*p ); moreover,
+            scalar multiplication and sum operations are available.
+
+        out : psydac.linalg.basic.Vector | NoneType
+            The output vector, or None (optional).
+
+        Results
+        -------
+        x : psydac.linalg.basic.Vector
+            Converged solution.
+
+        info : dict
+            Dictionary containing convergence information:
+            - 'niter'    = (int) number of iterations
+            - 'success'  = (boolean) whether convergence criteria have been met
+            - 'res_norm' = (float) 2-norm of residual vector r = A*x - b.
```

Comment:
The `solve` function updates `self._info`, but does not return the variable itself. I think this error in the docstring appears throughout the file.

If we want this function to return `info` we should change the last lines to
```python
        # Convergence information
        info = {'niter': k+1, 'success': am < tol, 'res_norm': am }

        return x, info
```
and then update the `dot` accordingly:
```python
    def dot(self, b, out=None):
        x, info = self.solve(b, out=out)
        self._info = info
        return x
```

@e-moral-sanchez @jowezarek @tomcaruso What do you think? Is it worth making this change? Or should we just update the docstrings of the `solve` methods instead?

Reply:
I'd vote against making this change. Not only because we'd have to change all tests (`x = A_inv @ b` -> `x, info = A_inv @ b`), but also because I wouldn't want to store convergence information by default but only when I'm interested in it.

Reply:
I agree with Julian. The .solve method should only return the solution and not the info. I think that it is less confusing, if we only change the docstring.

Reply:
All right. Let's change the docstrings then!

Reply:
Then where should I explain the Object `._info`?

Reply:
OK, I did the changes but I am not convinced, because the iterative solvers of Scipy return a `Tuple` of solution and info.

Reply:
> OK, I did the changes but I am not convinced, because the iterative solvers of Scipy return a `Tuple` of solution and info.

The iterative solvers of Scipy are simple functions, so they must return `info` when called. Our solvers are `InverseLinearOperator` objects and therefore they can be inspected at a later time. I do not see why we should mimic the Scipy interface now that we have proper classes instead of simple functions.

Personally I do not see the reason to expose the function `solve` to the user, as it is not part of the public interface of the `LinearOperator` objects. I would go as far as calling it `_solve` to make it a private function.